### PR TITLE
Refactoring retry logic.

### PR DESCRIPTION
**Pull request by agoulti**
We will need to reuse it for Download.

### CODE REVIEW

File: src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+   * HttpResponseException with the code Unauthorized is caught. Any other
+   * exceptions are passed through.
+   *
+   * @param a Action to execute
+   * @param retries How many times to attempt to refresh credentials if there
+   * is no progress. (Every time an action successfully completes, the retry
+   * budget is reset)
+   * @param <Ex> An action-specific exception that might be throwns.
+   */
+
```
Comment (bwkimmel) [inline src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java]: Nit: remove blank line.

File: src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jenkins.plugins.storage.util;
+
+import static java.util.logging.Level.SEVERE;
+
+import java.io.IOException;
+import java.util.LinkedList;
```
Comment (bwkimmel) [inline src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java]: Run mvn checkstyle:checkstyle.
(here: unused import)

File: src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ * A class to fascilitate retries on storage operations.
+ */
+public class RetryStorageOperation {
+  private static final Logger logger =
+      Logger.getLogger(RetryStorageOperation.class.getName());
+
+  /**
+   * An operation to be retried
+   */
+  public interface Action {
```
Comment (bwkimmel) [inline src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java]: The name "Action" is used in the Jenkins object model, which makes this usage a bit confusing. Maybe call these "Operation" and "RepeatableOperation" instead?

File: src/main/java/com/google/jenkins/plugins/storage/AbstractUpload.java [hunk 4]
```diff
@@ -418,83 +418,92 @@ public void checkRoles(RoleChecker checker)
...
+    RepeatAction<UploadException> a = new RepeatAction<UploadException>() {
+      private Queue<FilePath> paths;
+      Storage service;
+      Executor executor;
+      Bucket bucket;
+
+      {
+        paths = new LinkedList<>(uploads.inclusions);
+        executor = module.newExecutor();
+      }
```
[the remark this answers is not in the document]
Comment (agoulti) [reply src/main/java/com/google/jenkins/plugins/storage/AbstractUpload.java]: Oh cool, I didn't realize I could do that!

[the remark this answers is not in the document]
Comment (agoulti) [reply src/main/java/com/google/jenkins/plugins/storage/AbstractUpload.java]: Oh cool, I didn't realize I could do that!

File: src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java (hunk 3 above)
Comment (agoulti) [reply src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java]: Good point! Thanks for the suggestion!

File: src/test/java/com/google/jenkins/plugins/storage/util/RetryStorageOperationTest.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    public void act() throws IOException {
+      fails--;
+      if (fails >= 0) {
+        throw new IOException("bad");
+      }
+      succeeded++;
+    }
+  };
```
Comment (bwkimmel) [inline src/test/java/com/google/jenkins/plugins/storage/util/RetryStorageOperationTest.java]: Nit: Semicolons are not required after class definitions.

File: src/test/java/com/google/jenkins/plugins/storage/util/RetryStorageOperationTest.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        }
+        tries--;
+
+        if (tries % 2 == 1) {
+          throw new IOException("IOExcpetion");
+        }
+        throw new InterruptedException();
+      }
+    }
+    ;
```
Comment (bwkimmel) [inline src/test/java/com/google/jenkins/plugins/storage/util/RetryStorageOperationTest.java]: -;

**Pull request merged**